### PR TITLE
Issue 18 Colored Gauss Seidel

### PR DESCRIPTION
**Pull request by LinhWuerzburger**
update old code

### CODE REVIEW

**Comment by LinhWuerzburger** [comment]
rewrote CGS, now it compiles with pragmas again (#18)
and it excludes obstacle cells in its calculation (#164)

but the result of the GPU version is not correct, will be handled in #130

File: src/diffusion/ColoredGaussSeidelDiffuse.cpp [hunk 1]
```diff
@@ -133,78 +113,69 @@ void ColoredGaussSeidelDiffuse::diffuse(
...
+    size_t *domain_inner_list = domain_controller->get_domain_inner_list_level_joined();
 
-//#pragma acc data present(d_out[:bsize], d_b[:bsize], d_EV[:bsize])
-{
-    const size_t Nx = domain_data->get_Nx();
-    const size_t Ny = domain_data->get_Ny();
+#pragma acc data present(out, b, EV)
+    {
+        const size_t Nx = domain_data->get_Nx();
+        const size_t Ny = domain_data->get_Ny();
```
Comment (ChristianVonMach) [inline src/diffusion/ColoredGaussSeidelDiffuse.cpp:129]: those variables aren't used in this function

File: src/diffusion/ColoredGaussSeidelDiffuse.cpp [hunk 2]
```diff
@@ -45,80 +39,66 @@ ColoredGaussSeidelDiffuse::ColoredGaussSeidelDiffuse(Settings::Settings const &s
...
+            for (size_t j = 0; j < size_domain_inner_list; ++j) {
+                const size_t index = domain_inner_list[j];
+                res = reciprocal_beta * (out[index] - in[index]);
+                sum += res * res;
+            }
+
+#pragma acc wait
+            res = sqrt(sum);
+            it++;
+        } //end while
```
Comment (ChristianVonMach) [inline src/diffusion/ColoredGaussSeidelDiffuse.cpp:91]: old comment style

File: src/diffusion/ColoredGaussSeidelDiffuse.cpp [hunk 3]
```diff
@@ -45,80 +39,66 @@ ColoredGaussSeidelDiffuse::ColoredGaussSeidelDiffuse(Settings::Settings const &s
...
+        real sum;
+        real res = 1.;
+        while (res > m_tol_res && it < m_max_iter) {
+            in.copy_data(out);  // necessary for calculation of residuum
+            colored_gauss_seidel_step(out, b, alpha_x, alpha_y, alpha_z, beta, m_dsign, m_w, odd_indices, even_indices, sync);
+            domain_controller->apply_boundary(out, sync);
+
+            sum = 0;
+
+#pragma acc parallel loop independent present(out, in, domain_inner_list[:size_domain_inner_list]) async
```
Comment (ChristianVonMach) [inline src/diffusion/ColoredGaussSeidelDiffuse.cpp:81]: use reduction pragma

File: src/diffusion/ColoredGaussSeidelDiffuse.cpp [hunk 4]
```diff
@@ -133,78 +113,69 @@ void ColoredGaussSeidelDiffuse::diffuse(
...
-//#pragma acc wait
-    }
+        while (res > m_tol_res && it < m_max_iter) {
+            in.copy_data(out);  // necessary for calculation of residuum
+            colored_gauss_seidel_step(out, b, m_dsign, m_w, D, EV, dt, odd_indices, even_indices, sync);
+            domain_controller->apply_boundary(out, sync);
+
+            sum = 0;
+
+#pragma acc parallel loop independent present(out, b, EV, domain_inner_list[:size_domain_inner_list]) async
```
Comment (ChristianVonMach) [inline src/diffusion/ColoredGaussSeidelDiffuse.cpp:154]: use reduction pragma

**Review by ChristianVonMach — CHANGES_REQUESTED** [review]

File: src/diffusion/ColoredGaussSeidelDiffuse.cpp (hunk 1 above)
Comment (LinhWuerzburger) [reply src/diffusion/ColoredGaussSeidelDiffuse.cpp]: 8fba1a3ca4c5c48d015a448dda7f852212fbe8da

File: src/diffusion/ColoredGaussSeidelDiffuse.cpp (hunk 2 above)
Comment (LinhWuerzburger) [reply src/diffusion/ColoredGaussSeidelDiffuse.cpp]: a3f83f3

File: src/diffusion/ColoredGaussSeidelDiffuse.cpp (hunk 3 above)
Comment (LinhWuerzburger) [reply src/diffusion/ColoredGaussSeidelDiffuse.cpp]: d7d2b90

File: src/diffusion/ColoredGaussSeidelDiffuse.cpp (hunk 4 above)
Comment (LinhWuerzburger) [reply src/diffusion/ColoredGaussSeidelDiffuse.cpp]: d7d2b90

**Review by ChristianVonMach — DISMISSED** [review]

**Review by ChristianVonMach — APPROVED** [review]

**Pull request merged**